src/js/d3_svg_cell_renderer.js

```diff
@@ -0,0 +1,207 @@
+var utils = require('./utils');
+var $ = require('jquery');
+var _ = require('underscore');
+
+// TODO: handle accessing config properties cleaner
+
+function D3SVGRuleSet(renderer) {
+	var self = this;
+	self.rule_map = {};
+	self.renderer = renderer;
+
+	self.addRule = function(condition, d3_shape, attrs, z_index) {
+		var rule_id = Object.keys(self.rule_map).length;
+		if (z_index === undefined) {
+			z_index = rule_id;
+		}
+		self.rule_map[rule_id] = new D3SVGRule(rule_id, self.renderer, condition, d3_shape, attrs, z_index);
+		return rule_id;
+	};
+
+	self.removeRule = function(rule_id) {
+		delete self.rule_map[rule_id];
+	}
+
+	self.getRules = function() {
+		// returns a list of lists of rules to render in the given order
+		var z_map = {};
+		_.each(self.rule_map, function(rule, rule_id) {
+			z_map[rule.z_index] = z_map[rule.z_index] || [];
+			z_map[rule.z_index].push(rule);
+		});
+		return _.map(Object.keys(z_map).sort(), 
+			function(z) { return z_map[z];});
+	};
+
+	self.fromJSON = function(json_rules) {
+		self.rule_map = {};
+		_.each(json_rules, function(rule) {
+			self.addRule(rule.condition, rule.d3_shape, rule.attrs, rule.z_index);
+		});
+	};
+}
+
+function D3SVGRule(rule_id, renderer, condition, d3_shape, attrs, z_index) {
```

Comment:
1. If you look at the code [here](https://github.com/cBioPortal/oncoprintjs/blob/master/src/js/main.js#L81) you'll see the implementation of a fluent interface which is self-documenting. Validation can be done via some sort of helper function.

2. I hear you saying that you want name spacing to organize methods. This indeed one of ways that classes are used in others languages (again, like Java). Fortunately, we don't need classes to group functions together because we have nodejs' `require` system and browerify to link all our files together. Name spacing and grouping together of functions can be done by file. This allows us to avoid having to unnecessarily maintain object state. Of course, be there a situation where state is truly necessary, we can always create an object to handle that state.

3. I don't follow. I still don't understand why the code has to be restructured in this particular way in order to support the features that we want: abstraction from SVG based renderering. This has been my question from the beginning and I am asking it again now.